core/src/main/scala/org/apache/spark/executor/ProcfsMetricsGetter.scala

```diff
@@ -138,19 +138,22 @@ private[spark] class ProcfsMetricsGetter(
       }
       val stdoutThread = Utils.processStreamByLine("read stdout for pgrep",
         process.getInputStream, appendChildPid)
-      val error = process.getErrorStream
-      var errorString = ""
-      (0 until error.available()).foreach { i =>
-        errorString += error.read()
-      }
+      val errorStringBuilder = new StringBuilder()
+      val stdErrThread = Utils.processStreamByLine(
+        "stderr for pgrep",
+        process.getErrorStream,
+        { line =>
+        errorStringBuilder.append(line)
+      })
```

Comment:
nit: if the line-handling closure is multiline, you should indent the body more.  but I'd just put it on one line

```scala
val stdErrThread = Utils.processStreamByLine(
  "stderr for pgrep",
  process.getErrorStream,
  line => errorStringBuilder.append(line))
```